### DISCUSSION TITLE
Genius Lyrics Lookup

### DISCUSSION THREAD

**Discussion by dealingwith** [johnlindquist/kit#392] [Share]
Just what it says on the tin. How can I improve this one?

[Install song-search](https://scriptkit.com/api/new?name=song-search&url=https://gist.githubusercontent.com/dealingwith/11ca5e660d3873a3e521ad2905e2b5b7/raw/4910306cd3818772b1cd67b1718e0f3d59e7ced5/song-search.js)

```js
// Menu: Genius Lyrics Search

let Genius = await npm('genius-lyrics-api')

let geniusUserToken = await env("GENIUS_AUTH_TOKEN")
import { getLyrics, searchSong } from 'genius-lyrics-api';

let songTitle = await arg("Song Title")

let options = {
  apiKey: geniusUserToken,
  title: songTitle,
  artist: '',
  optimizeQuery: true
}

let returnedSongs = await(searchSong(options).then((r) => r))
let returnedSongTitles = returnedSongs.map(s => s.title)

let chosenSongTitle = await arg("Which song?", returnedSongTitles)

let songData = returnedSongs.filter(s => s.title == chosenSongTitle)
let lyrics = await(getLyrics(songData[0].url).then((r) => r))

let whatToDo = await arg("", ["Copy", "View", "Visit"])

if (whatToDo == "Copy") copy(lyrics)
else if (whatToDo == "View") await textarea(lyrics)
else if (whatToDo == "Visit") await focusTab(songData[0].url)

```

**Comment by dealingwith** [comment]
Is there a better component to use than `textarea` for displaying plain text? Should `esc` be able to dismiss a textarea control? The only way I seem to be able to get out of it is `cmd+s`.

**Reply by johnlindquist** [reply]
You can dismiss a `textarea` with `cmd+w`. The idea is that it mimics the shortcuts of the `editor` component and there are many instances when you would want to use `escape` in the editor. Maybe I should rethink that?

**Comment by johnlindquist** [comment]
Really nice work, love it. I didn't realize Genius had an api.

Added some minor changes to the script:

1. you can destructure off of `await npm`. If you `import` like it was before, then the script tries to import before anything runs.
2. I set the `env` to be a "password" so it shows the ***** instead of the plain text. I'm adding a `type` so it will support all the different text-based input types "email", "url", "password", etc
3. I added the album cover as an `img`

```js
// Menu: Genius Lyrics Search
// Description: Search for lyrics
// Author: Daniel Miller
// Twitter: @danielindustri4

let { getLyrics, searchSong } = await npm(
  "genius-lyrics-api"
)

let geniusUserToken = await env("GENIUS_AUTH_TOKEN", {
  secret: "password",
})

let songTitle = await arg("Song Title")

let options = {
  apiKey: geniusUserToken,
  title: songTitle,
  artist: "",
  optimizeQuery: true,
}

let returnedSongs = await searchSong(options).then(r => r)
let returnedSongTitles = returnedSongs.map(s => s.title)

let chosenSongTitle = await arg(
  "Which song?",
  returnedSongTitles
)

let songData = returnedSongs.filter(
  s => s.title == chosenSongTitle
)

let song = songData[0]

let lyrics = await getLyrics(song.url).then(r => r)

let whatToDo = await arg(
  {
    placeholder: "What do you want to do?",
    selected: song.title,
  },
  [
    {
      name: "Copy",
      value: "Copy",
      description: "Copy to clipboard",
    },
    {
      name: "View",
      value: "View",
      description: "Open lyrics in a TextArea",
    },
    {
      name: "Visit",
      value: "Visit",
      description: songData[0].url,
      img: song.albumArt,
    },
  ]
)

if (whatToDo == "Copy") copy(lyrics)
else if (whatToDo == "View") await textarea(lyrics)
else if (whatToDo == "Visit") await focusTab(song.url)

```

**Reply by dealingwith** [reply]
Thanks for the upgrade and the tips!

![awesome](https://media0.giphy.com/media/hZfm9Pj95F9Mk/giphy.gif?cid=ba572133wkbfg0lxuv1updqxa5lec6oefjxfunxmji4t3h7z&rid=giphy.gif&ct=g)

**Reply by dealingwith** [reply]
oh and it's danielindustri4 on Twitter 😉
